miner/block_builder.go

```diff
@@ -177,30 +178,60 @@ func calcHdistRange(id types.LayerID, hdist types.LayerID) (bottom types.LayerID
 	return bottom, top
 }
 
-func (t *BlockBuilder) createBlock(id types.LayerID, atxID types.AtxId, eligibilityProof types.BlockEligibilityProof,
-	txids []types.TransactionId, atxids []types.AtxId) (*types.Block, error) {
-
+func (t *BlockBuilder) getVotes(id types.LayerID) ([]types.BlockID, error) {
 	var votes []types.BlockID = nil
-	var err error
+
+	// if genesis
 	if id == config.Genesis {
 		return nil, errors.New("cannot create blockBytes in genesis layer")
-	} else if id == config.Genesis+1 {
-		votes = append(votes, config.GenesisId)
-	} else { // get from hare
-		bottom, top := calcHdistRange(id, t.hdist)
-		votes, err = t.hareResult.GetResult(bottom, top)
-		if err != nil {
-			t.With().Warning("Could not get hare result during block creation",
-				log.Uint64("bottom", uint64(bottom)), log.Uint64("top", uint64(top)),
-				log.Uint64("hdist", uint64(t.hdist)), log.Err(err))
+	}
+
+	// if genesis+1
+	if id == config.Genesis+1 {
+		return append(votes, config.GenesisId), nil
+	}
+
+	// not genesis, get from hare
+	bottom, top := calcHdistRange(id, t.hdist)
+
+	votesBottom, err := t.hareResult.GetResult(bottom, bottom)
+	if votesBottom == nil || len(votesBottom) == 0 {
+		t.With().Info("bottom votes is nil or empty. adding the whole layer", log.LayerId(uint64(bottom)))
+		l, e := t.meshProvider.GetLayer(bottom)
+		if e != nil {
+			t.With().Error("Could not set votes to whole layer", log.Err(e))
+			return nil, e
 		}
-		if votes == nil { // if no votes set to empty
-			t.Info("Votes is nil. Setting votes to an empty array")
-			votes = []types.BlockID{}
+
+		// set votes to whole layer
+		for _, b := range l.Blocks() {
```

Comment:
why not use rangeVotes, err := t.hareResult.GetResult(bottom, top) once and if bottom nil catch and error and add the whole layer ?
not critical but this code is a little convoluted

Reply:
better even to change GetResult so it will work per layer and do the whole logic here

Reply:
Done